incubator/virtualcluster/pkg/syncer/controllers/pod/checker.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pod
+
+import (
+	"context"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+
+	ctrl "github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/controller"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/conversion"
+)
+
+// StartPeriodChecker starts the period checker for data consistency check. Checker is
+// blocking so should be called via a goroutine.
+func (c *controller) StartPeriodChecker(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Start a loop to periodically check if pods keep consistency between super
+	// master and tenant masters.
+	wait.Until(c.checkPods, c.periodCheckerPeriod, stopCh)
+}
+
+// checkPods checks to see if pods in super master informer cache and tenant master
+// keep consistency.
+func (c *controller) checkPods() {
+	clusterNames := c.multiClusterPodController.GetClusterNames()
+	if len(clusterNames) == 0 {
+		klog.Infof("tenant masters has no clusters, give up period checker")
+		return
+	}
+
+	for _, clusterName := range clusterNames {
+		cluster := c.multiClusterPodController.GetCluster(clusterName)
+		if cluster == nil {
+			klog.Errorf("failed to locate cluster %s", clusterName)
+			continue
+		}
+
+		c.checkPodsOfCluster(clusterName, cluster)
+	}
+
+	pPods, err := c.podLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("error listing pods from super master informer cache: %v", err)
+		return
+	}
+
+	for _, pPod := range pPods {
+		clusterName, vNamespace := conversion.GetOwner(pPod)
+		vPodObj, err := c.multiClusterPodController.Get(clusterName, vNamespace, pPod.Name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				if pPod.DeletionTimestamp != nil {
+					// pPod is under deletion, waiting for UWS bock populate the pod status.
+					continue
+				}
+				if err = c.client.Pods(pPod.Namespace).Delete(pPod.Name, &metav1.DeleteOptions{}); err != nil {
```

Comment:
Is there a chance that during syncer restart, the vcluster has not been added to multiClusterPodController yet but we reach here? Then the Pods may be wrongly deleted.